ee/clickhouse/sql/dead_letter_queue.py

```diff
@@ -0,0 +1,108 @@
+from ee.clickhouse.sql.clickhouse import KAFKA_COLUMNS, REPLACING_MERGE_TREE, kafka_engine, table_engine, ttl_period
+from ee.kafka_client.topics import KAFKA_DEAD_LETTER_QUEUE
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+
+# We pipe our Kafka dead letter queue into CH for easier analysis and longer retention
```

Comment:
If we consider Kafka as the _authoritative source of raw events_ I don’t think we should pipe and duplicate this data in ClickHouse as well. 

If the argument is that we want to have a longer retention period we should probably just configure the new Kafka topic to use a higher [retention.bytes](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#topicconfigs_retention.bytes) or [retention.ms](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#topicconfigs_retention.ms) value. I see the point of easier analysis in CH, but we can probably work around that by producing better metrics in the ingestion app.

Last additional note: from a data integrity perspective, trying to synchronise a dataset across two different datastore might lead to higher chances of having data out of sync (as well as duplicating storage costs).

Let me know what do you think about it 🙇

Reply:
I wouldn't feel super comfortable storing this for too long in Kafka + processing data is much easier in CH.

For instance, with the last issue we had, we got everything out of Kafka and then went through quite some pain to sort the data and get it in the right format. So it's not just about metrics per se, but also other operations we might want to run over the entire dataset. 

With Kafka we have to consume it all first and have it somewhere before being able to do anything about it. 

As for costs, ultimately this table really shouldn't store much, although of course that's not something to rely on. With the outages we had recently that hadn't happened for a while (e.g. Heroku Postgres going down) we would have had 4.5M or so rows here to be kept for a few months max.

Nevertheless, you're more knowledgeable than me here, so happy to get pushback. I also think this approach allows us to move fast though while we work on improving metrics, reliability, etc.

This did make me think about storage though, and realistically I think this could probably have a lower TTL. A month maybe. That should be more than enough time to make decisions on dropping/replaying events.

Reply:
Let's go on with this implementation so. We can always revisit it in the future.